### PR TITLE
US26069 : Documentation template for load balancer

### PR DESCRIPTION
**Pull request by Pushpalatha555**
---
layout: ""
page_title: "hpegl_vmaas_load_balancer Resource - vmaas-terraform-resources"
subcategory: "vmaas"
description: |-
    loadbalancer resource facilitates creating, updating
          and deleting NSX-T  Network Load Balancers.
---

# Resource hpegl_vmaas_load_balancer

-> Compatible version >= 5.4.6

loadbalancer resource facilitates creating, updating
		and deleting NSX-T  Network Load Balancers.
`hpegl_vmaas_load_balancer` resource supports NSX-T Load Balancer creation.

## Example usage for creating NSX-T Load Balancer with all possible attributes

```terraform
# (C) Copyright 2022 Hewlett Packard Enterprise Development LP

resource "hpegl_vmaas_load_balancer" "tf_lb" {
  name        = "tf_loadbalancer"
  description = "Loadbalancer created using tf"
  enabled     = true
  group_access {
    all = true
  }
  config {
    admin_state    = true
    size           = "SMALL"
    log_level      = "INFO"
    tier1_gateways = data.hpegl_vmaas_router.tier1_router.provider_id
  }
}
```

### Required

- `config` (Block List, Min: 1) Network Load Balancer Configuration (see [below for nested schema](#nestedblock--config))
- `name` (String) Network loadbalancer name

### Optional

- `description` (String) Creating the  Network loadbalancer
- `enabled` (Boolean) Pass `true` to allow for enabled and Pass `false` to disabled
- `group_access` (Block List, Max: 1) (see [below for nested schema](#nestedblock--group_access))

### Read-Only

- `id` (String) The ID of this resource.
- `lb_type` (String) Type of Network loadbalancer
- `network_server_id` (Number) NSX-T Integration ID

<a id="nestedblock--config"></a>
### Nested Schema for `config`

Required:

- `tier1_gateways` (String) Provider ID of the Tier1 Gateway. Use hpegl_vmaas_router datasource to obtain the provider_id here.

Optional:

- `admin_state` (Boolean) If `true` then admin State rule will be active/enabled.
- `log_level` (String) In Filter. Supported Values are "DEBUG", "INFO", "WARNING", "ERROR", "CRITICAL", "ALERT", "EMERGENCY"
- `size` (String) In Filter. Supported Values are "SMALL", "MEDIUM", "LARGE"

<a id="nestedblock--group_access"></a>
### Nested Schema for `group_access`

Optional:

- `all` (Boolean) Pass `true` to allow access to all groups.
- `sites` (Block List) List of sites/groups (see [below for nested schema](#nestedblock--group_access--sites))

<a id="nestedblock--group_access--sites"></a>
### Nested Schema for `group_access.sites`

Optional:

- `default` (Boolean) Group Default Selection
- `id` (Number) ID of the site/group

### CODE REVIEW

File: internal/resources/resource_loadbalancer.go [hunk 1]
```diff
@@ -101,9 +101,10 @@ func LoadBalancer() *schema.Resource {
 							Description: `In Filter. Supported Values are "DEBUG", "INFO", "WARNING", "ERROR", "CRITICAL", "ALERT", "EMERGENCY"`,
 						},
 						"tier1_gateways": {
-							Type:        schema.TypeString,
-							Required:    true,
-							Description: "Provider ID of the Tier1 Gateway. Use " + DSRouter + " datasource to obtain the provider_id  here.",
+							Type:     schema.TypeString,
+							Required: true,
+							Description: "Tier1 Gateway. Get the `providerId`. Use " + DSRouter +
+								"datasource to obtain the provider_id and nsx-t-tier1-gateway which indicates that it is a NSX T Tier1 Gateway",
```
Comment (manjunath-batakurki) [inline internal/resources/resource_loadbalancer.go:107]: NSX-T

File: templates/resources/vmaas_load_balancer.md.tmpl [hunk 2]
```diff
@@ -0,0 +1,25 @@
...
+
+-> Compatible version >= 5.4.6
+
+{{ .Description | trimspace }}
+`hpegl_vmaas_load_balancer` resource supports NSX-T Load Balancer creation.
+
+-> It is mandatory to choose the `Tier1 Gateway` while creating the load balancer
+(for this purpose `hpegl_vmaas_router` can be used).
+
+-> Tier1 Gateway Data Source `hpegl_vmaas_router` is supported for NSX-T Load Balancer.
```
Comment (manjunath-batakurki) [inline templates/resources/vmaas_load_balancer.md.tmpl:19]: Tier1 Gateway Data Source `hpegl_vmaas_router` is supported for NSX-T Load Balancer is not needed.

Comment (Pushpalatha555) [reply templates/resources/vmaas_load_balancer.md.tmpl]: removed

File: internal/resources/resource_loadbalancer.go [hunk 3]
```diff
@@ -101,9 +101,10 @@ func LoadBalancer() *schema.Resource {
 							Description: `In Filter. Supported Values are "DEBUG", "INFO", "WARNING", "ERROR", "CRITICAL", "ALERT", "EMERGENCY"`,
 						},
 						"tier1_gateways": {
-							Type:        schema.TypeString,
-							Required:    true,
-							Description: "Provider ID of the Tier1 Gateway. Use " + DSRouter + " datasource to obtain the provider_id  here.",
+							Type:     schema.TypeString,
+							Required: true,
+							Description: "Tier1 Gateway. Get the `providerId`. Use " + DSRouter +
+								"datasource to obtain the provider_id and nsx-t-tier1-gateway which indicates that it is a NSX-T Tier1 Gateway",
```
Comment (manjunath-batakurki) [inline internal/resources/resource_loadbalancer.go:107]: I guess this is better and simple. "Provider ID of the Tier1 Gateway. Use " + DSRouter + " datasource to obtain the `provider_id` here.",

File: templates/resources/vmaas_load_balancer.md.tmpl [hunk 4]
```diff
@@ -0,0 +1,23 @@
...
+---
+
+# Resource hpegl_vmaas_load_balancer
+
+-> Compatible version >= 5.4.6
+
+{{ .Description | trimspace }}
+`hpegl_vmaas_load_balancer` resource supports NSX-T Load Balancer creation.
+
+-> It is mandatory to choose the `Tier1 Gateway` while creating the load balancer
```
Comment (manjunath-batakurki) [inline templates/resources/vmaas_load_balancer.md.tmpl:16]: why do you need talk about Tier1 ? does the schema doesnt say its a required field ?

Comment (Pushpalatha555) [reply templates/resources/vmaas_load_balancer.md.tmpl:16]: here it is not required but for creating the virtual server from LB without Tier1 will fail to create so the precaution added here only

I will remove it if not needed

File: internal/resources/resource_loadbalancer.go (hunk 3 above)
Comment (Pushpalatha555) [reply internal/resources/resource_loadbalancer.go:107]: updated

**Review by manjunath-batakurki — APPROVED** [review]

**Review by ravikumar29524 — APPROVED** [review]

**Pull request merged**